### PR TITLE
Fix plots definition union issues for VS Code

### LINKED ISSUE #33: Some plots definitions appear broken
Came across some weirdness when getting back to our LSP. Seems like there are some issues with the schema for top-level plots:

image

image

image

Not all definitions are broken:

image

image

### PR DESCRIPTION
**Pull request by mattseddon**
Fixes #33

Comments inline.

### Before

https://user-images.githubusercontent.com/37993418/215945282-4ebecf29-d784-4d77-840c-1b3a11091e2d.mov

### After

https://user-images.githubusercontent.com/37993418/215945083-d8c85453-4959-4ab7-89f2-8dac0266e9a1.mov

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -35,7 +35,7 @@ repos:
       - id: pyupgrade
         args: ['--py310-plus']
   - repo: https://github.com/PyCQA/isort
-    rev: 5.10.1
+    rev: 5.12.0
```
Comment (mattseddon) [inline .pre-commit-config.yaml:38]: [F] I couldn't get the pre-commit to run without making this upgrade. I am happy to pull it out into a separate PR if needed 🙏🏻.

File: gen.py [hunk 2]
```diff
@@ -72,15 +72,25 @@ class PlotFlags(OutFlags):
...
+
+
+# pylint: disable=invalid-name
+class Y(BaseModel):
+    __root__: dict[FilePath, PlotColumn | PlotColumns] = Field(None)
+
+
 class TopLevelPlotFlags(BaseModel):
-    x: PlotColumn | dict[FilePath, PlotColumn] = Field(
+    x: PlotColumn | X = Field(
```
Comment (mattseddon) [inline gen.py:86]: [F] In order to get the Red Hat Yaml extension to de-duplicate the entries for x/y and top-level plots I had to split out the types like this. I have never used Pydantic before so please correct me if there is a better way to do this.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (skshetry) [reply .pre-commit-config.yaml:38]: That's fine. Thank you for updating it.

File: gen.py [hunk 3]
```diff
@@ -72,15 +72,25 @@ class PlotFlags(OutFlags):
     template: FilePath = Field(None, description="Default plot template")
 
 
+# pylint: disable=invalid-name
+class X(BaseModel):
```
Comment (skshetry) [inline gen.py:76]: Let's name it since this is part of only what `x` supports. I don't have good names either to suggest, how about `PlotFileColumnMap`/`PlotFileColumnDict` etc? If you can think of good names, please feel free.

File: gen.py [hunk 4]
```diff
@@ -72,15 +72,25 @@ class PlotFlags(OutFlags):
...
+    __root__: dict[FilePath, PlotColumn] = Field(None)
```
Comment (skshetry) [inline gen.py:77]:
```suggestion
    __root__: dict[FilePath, PlotColumn] = Field(default_factory=dict)
```

File: gen.py [hunk 5]
```diff
@@ -72,15 +72,25 @@ class PlotFlags(OutFlags):
...
+
+
+# pylint: disable=invalid-name
+class Y(BaseModel):
+    __root__: dict[FilePath, PlotColumn | PlotColumns] = Field(None)
```
Comment (skshetry) [inline gen.py:82]:
```suggestion
    __root__: dict[FilePath, PlotColumn | PlotColumns] = Field(default_factory=dict)
```

File: gen.py [hunk 6]
```diff
@@ -72,15 +72,25 @@ class PlotFlags(OutFlags):
     template: FilePath = Field(None, description="Default plot template")
 
 
+# pylint: disable=invalid-name
+class X(BaseModel):
+    __root__: dict[FilePath, PlotColumn] = Field(None)
+
+
+# pylint: disable=invalid-name
+class Y(BaseModel):
```
Comment (skshetry) [inline gen.py:81]: Also, how about `PlotFileColumnsMap`/`PlotFileColumnsDict` as above?

File: gen.py (hunk 5 above)
Comment (skshetry) [inline gen.py:82]: I know the types are confusing, but `PlotColumns` is union of `PlotColumn | set[PlotColumn]`, so we don't need a union here.
```suggestion
    __root__: dict[FilePath, PlotColumns] = Field(None)
```

File: gen.py [hunk 7]
```diff
@@ -312,8 +322,14 @@ class Config:
 Use elsewhere in `dvc.yaml` with the `${}` substitution expression."""
 
 
-PlotsDict = dict[PlotIdOrFilePath, TopLevelPlotFlags | EmptyTopLevelPlotFlags]
-PlotsList = list[PlotIdOrFilePath | PlotsDict]
+class TopLevelPlots(BaseModel):
+    __root__: dict[
+        PlotIdOrFilePath, TopLevelPlotFlags | EmptyTopLevelPlotFlags
+    ] = Field(None)
```
Comment (skshetry) [inline gen.py:328]:
```suggestion
    ] = Field(default_factory=dict)
```

File: gen.py [hunk 8]
```diff
@@ -312,8 +322,14 @@ class Config:
...
+
+
+class TopLevelPlotsList(BaseModel):
+    __root__: list[PlotIdOrFilePath | TopLevelPlots] = Field(None)
```
Comment (skshetry) [inline gen.py:332]:
```suggestion
    __root__: list[PlotIdOrFilePath | TopLevelPlots] = Field(default_factory=list)
```

File: gen.py (hunk 2 above)
Comment (skshetry) [reply gen.py:86]: I could not find an issue in pydantic as well. Probably because this issue is more about what to render in the extension side.

**Comment by skshetry** [comment]
@mattseddon, looks good overall. I have a few minor suggestions inlined.

File: gen.py (hunk 6 above)
Comment (mattseddon) [reply gen.py:81]: The only reason that I used `X`/`Y` for names is that they show up as the title of the field in the `onHover` provider. E.g if I change the `X` class to `PlotFileColumnsDict` we get:

image


WDYT?

**Comment by mattseddon** [comment]
@skshetry, thank you for your review 🙏🏻. I applied two of your suggestions. The third one I have left as is for the time being (didn't rename X/Y for the reason given) but happy to be overruled if you feel strongly about it. Thanks again.

File: gen.py (hunk 6 above)
Comment (skshetry) [reply gen.py:81]: I was thinking of adding `title=""` to `Field`, but that does not seem to work for `__root__`.

**Review by skshetry — APPROVED** [review]

**Comment by mattseddon** [comment]
Thanks
